lib/exgboost/training/callback.ex

```diff
@@ -134,54 +140,34 @@ defmodule EXGBoost.Training.Callback do
             "target metric #{inspect(target_metric)} not found in metrics #{inspect(metrics)}"
     end
 
-    prev_criteria_value = best
-
-    cur_criteria_value = metrics[target_eval][target_metric]
+    score = metrics[target_eval][target_metric]
 
     improved? =
-      case mode do
-        :min ->
-          prev_criteria_value == nil or
-            cur_criteria_value < prev_criteria_value
-
-        :max ->
-          prev_criteria_value == nil or
-            cur_criteria_value > prev_criteria_value
+      cond do
+        best_score == nil -> true
+        mode == :min -> score < best_score
+        mode == :max -> score > best_score
       end
 
-    over_patience? = since_last_improvement >= patience
-
     cond do
       improved? ->
-        updated_meta_vars =
-          meta_vars
-          |> put_in([:early_stop, :best], cur_criteria_value)
-          |> put_in([:early_stop, :since_last_improvement], 0)
+        early_stop = %{early_stop | best: score, since_last_improvement: 0}
 
         bst =
           bst
-          |> struct(best_iteration: state.iteration, best_score: cur_criteria_value)
-          |> EXGBoost.Booster.set_attr(
-            best_iteration: state.iteration,
-            best_score: cur_criteria_value
-          )
+          |> struct(best_iteration: state.iteration, best_score: score)
+          |> EXGBoost.Booster.set_attr(best_iteration: state.iteration, best_score: score)
 
-        {:cont, %{state | meta_vars: updated_meta_vars, booster: bst}}
+        %{state | booster: bst, meta_vars: %{meta_vars | early_stop: early_stop}}
 
-      not improved? and not over_patience? ->
-        updated_meta_vars =
-          meta_vars
-          |> put_in([:early_stop, :since_last_improvement], since_last_improvement + 1)
-
-        {:cont, %{state | meta_vars: updated_meta_vars}}
+      since_last_improvement < patience ->
+        early_stop = Map.update!(early_stop, :since_last_improvement, &(&1 + 1))
+        %{state | meta_vars: %{meta_vars | early_stop: early_stop}}
 
       true ->
-        updated_meta_vars =
-          meta_vars
-          |> put_in([:early_stop, :since_last_improvement], since_last_improvement + 1)
-
-        bst = struct(bst, best_iteration: state.iteration, best_score: cur_criteria_value)
-        {:halt, %{state | meta_vars: updated_meta_vars, booster: bst}}
+        early_stop = Map.update!(early_stop, :since_last_improvement, &(&1 + 1))
+        bst = struct(bst, best_iteration: state.iteration, best_score: score)
```

Comment:
Is this possibly a bug? It doesn't match the other `cond` branch where:

```elixir
bst =
  bst
  |> struct(best_iteration: state.iteration, best_score: score)
  |> EXGBoost.Booster.set_attr(best_iteration: state.iteration, best_score: score)
```

I haven't had time time to dig into the particulars of what's going on in `Booster` yet, so I wasn't sure.

Reply:
Yeah it should probably either do bot add to struct and update the attirbute or just update the struct. The Booster attributes are held in the actual booster so are actually mutable. The implementation to use attributes exists but I dont think should really be used outside of maybe seeing any attributes from a serialized model that maybe were set elsewhere

Reply:
Also Im glad you cleaned up a lot of this code because IDK why I was using `put_in` so much